llvm-general/src/LLVM/General/OrcJIT.hs

```diff
@@ -12,7 +12,12 @@ module LLVM.General.OrcJIT (
     mangleSymbol,
     withIRCompileLayer,
     withModuleSet,
-    withObjectLinkingLayer
+    withObjectLinkingLayer,
+    addModuleSet,
+    createObjectLinkingLayer,
```

Comment:
Try implementing https://github.com/llvm-mirror/llvm/blob/master/examples/Kaleidoscope/BuildingAJIT/Chapter4/KaleidoscopeJIT.h#L172 with the API that’s currently exposed. In this case `addModuleSet` is not even enough you also need something like `moduleFromAST` without the `with` abstraction (which you didn’t want to have exposed either). We don’t have bindings to the API necessary for that example yet but exposing it now instead of when we have these bindings doesn’t do any additional harm.

In general `with` abstractions are simply not good enough for all usescases:

* In a callback they are mostly useless if you want the thing to live longer than the callback. Similar situations occur if you start doing multithreaded stuff.
* stack-based resource allocation is just one way to do resource allocation, there are other ways (e.g. `resourcet`) and `llvm-general` shouldn’t prevent users from using them.

You can probably hack around these issues in some ways but it gets more complicated and also not safer than the current API.

There is a reason why we have `openFile` and not just `withFile` and the same holds for the abstractions in `llvm-general`. We could just expose the internal modules (which I think is a good idea in general) but tbh I don’t think these things belong in internal modules: Internal modules are for things where you want to break the API arbitrarely and where the users can violate invariants of your data structures and imho neither of those applies here.